lib/spack/spack/filesystem_view.py

```diff
@@ -274,29 +265,96 @@ def add_standalone(self, spec):
                                         long=False)
                 return False
 
-        tree = LinkTree(spec.prefix)
-
-        if not self.ignore_conflicts:
-            conflict = tree.find_conflict(self.root)
-            if conflict is not None:
-                tty.error(self._croot +
-                          "Cannot link package %s, file already exists: %s"
-                          % (spec.name, conflict))
-                return False
+        self.merge(spec)
 
-        conflicts = tree.merge(self.root, link=self.link,
-                               ignore=ignore_metadata_dir,
-                               ignore_conflicts=self.ignore_conflicts)
         self.link_meta_folder(spec)
 
-        if self.ignore_conflicts:
-            for c in conflicts:
-                tty.warn(self._croot + "Could not link: %s" % c)
-
         if self.verbose:
             tty.info(self._croot + 'Linked package: %s' % colorize_spec(spec))
         return True
 
+    def merge(self, spec, ignore=None):
+        pkg = spec.package
+        if hasattr(pkg, 'view_source'):
+            view_source = pkg.view_source()
+        else:
+            view_source = spec.prefix
+        if hasattr(pkg, 'view_destination'):
+            view_dst = pkg.view_destination(self)
+        else:
+            view_dst = self.root
+
+        tree = LinkTree(view_source)
+
+        ignore = ignore or (lambda f: False)
+        ignore_file = match_predicate(
+            self.layout.hidden_file_paths, ignore)
+
+        # check for dir conflicts
+        conflicts = tree.find_dir_conflicts(view_dst, ignore_file)
+
+        # check for file conflicts (if the package reports 'get_file_conflicts'
+        # then use that, otherwise just check if any of the dst files exist)
+        merge_map = tree.get_file_map(view_dst, ignore_file)
+        if not self.ignore_conflicts:
+            if hasattr(pkg, 'view_file_conflicts'):
```

Comment:
Same here -- can we not use `hasattr` and either add the `view_file_conflicts` function to PackageBase, or add it to a `PackageBase` mixin class?